### DISCUSSION TITLE
Don't understand run_javascript

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#2044] [Q&A]
## Problem
I don't understand run_javascript.  I was trying to give focus to an input element when a dialog is opened.  I finally got it to work but 1) my solution is not documented anywhere and 2) I don't understand why it works.  Any info appreciated.

> [!NOTE]
> ui.run_javascript doc page is incomplete.  The _respond_ parameter is not documented there, which was crucial to making this work.

## Code
I borrowed focus() code from @falkoschindler [here](https://github.com/zauberzeug/nicegui/discussions/1218#discussioncomment-6511278).  Here's the first non-working version.  Dialog as in my previous example #2043:
```py
# --- create dialog
def rename_dialog (filename) :

    with ui.dialog () as dialog , ui.card ().classes ('dialog rename') :

        # input for new filename
        textin = ui.input (value = filename)

        # -----> give focus to input element ; this doesn't work <-----
        ui.run_javascript (f'''
            console.log ('foo') ;
            const element = getElement ({textin.id}).$refs.qRef.getNativeElement () ;
            element.focus () ;
        ''')

    dialog.open ()

# --- create page
@ui.page ('/image/{filename}') :
async def imagepage (filename, client : nicegui.Client) :
    await client.connected (10)
    # ... do other stuff ...
    rename_dialog ()
```

## Result
In code above, nothing happens.  Dialog opens as normal, input element is not focused.  There are no error messages in terminal or browser console.  Even 'foo' never shows up in browser console.  client is connected by the time rename_dialog runs.  Mysteriously, nothing happens.

## Next Attempt
Next I tried with await / async as in falko's example.  await on run_javascript, async on def rename_dialog.  That made things even worse.  Dialog never opens when async is added to method.  All I get is this message in terminal:
```
/Users/ed2050/dev/myscript.py:66: RuntimeWarning: coroutine 'rename_dialog' was never awaited
  rename_dialog (filename)
```

Of course it was never awaited.  I don't want any return value.  Just fire and forget.

## Almost there

Then I noticed the undocumented `respond = false` param in falko's example.  I added that to my code.  Again no dialog appears.  Now I get a similar error mesage:
```
/Users/ed2050/dev/myscript.py:96: RuntimeWarning: coroutine 'run_javascript' was never awaited
  ui.run_javascript (f'''
```

Of course it was never bloody awaited.  I told you not to await anything.  Isn't that the point of respond = false?

## Working solution

Then I added async on every method def in the chain, and await on every method call.  Which is quite annoying (the chain is longer than simplified code above).  I don't want or need all my methods to be async.  When I call an async method without await, it means just go away and do it, don't bother me with returns and stuff.

Anyway, just to humor the stupid interpreter I did it.  await / async everywhere.  I did my Oprah Winfrey impression: you get async, you get async, you get async!  And it finally worked.  I still have no idea why.

## Question

Why does every damn method need await / async on it just because of one stupid call to run_javascript?  Which I told not to bother returning a value anyway.

Ok I'm new to async.  I know the ins and outs of os.fork (), multiprocessing, threads, etc.  I read up on async, the model makes sense.  But I don't get why it throws roadblocks in my code.  Especially when the javascript call isn't even running in my server process.  Ship that code off to the browser and leave me alone.

**Comment by falkoschindler** [comment]
Which NiceGUI version are you working with?

The behavior you're describing is exactly what we meant to avoid with PR https://github.com/zauberzeug/nicegui/pull/1797. Until then there was the `respond` parameter to tell the client whether we need a response or not. Either way you needed to await `run_javascript()`, because the function was async (in case there is a response to be awaited). And that's how async functions work in Python: You need to await them (or put them in a new task and add them to the event look, see `background_tasks.create()`), otherwise they won't run.

The new `run_javascript()` doesn't have to be awaited. If you don't, it will fire and forget. It will return an `AwaitableResponse` which you can ignore. But if you await it, the client will know and respond accordingly.

We left the `respond` parameter to trigger a warning if it is used in old user code. But it is undocumented because you're not supposed to use it anymore.

**Reply by ed2050** [reply]
I'm using 1.3.16.  Great to know it was changed in a recent update.  I can cope for now.

How can I access docs for my version?  nicegui.io docs will be more and more out of date for me over time.

## Aside

Next part isn't about nicegui...
`<rant target="python-lang">` 
My problem with await is that using it forces you to change your method declaration.  And then any calls to your method must be awaited.  Then the calling methods have to change their declarations...  repeat ad nauseum.  Any model that forces silly cascading code changes is fundamentally broken.

If I have a method that starts multiple processes or threads, callers don't need to know or care.  What's next, making python adopt the java `throws` monstrosity where every method declares every conceivable exception in the known universe?  It's bad design.
`</rant>`

**Reply by rodja** [reply]
> How can I access docs for my version? nicegui.io docs will be more and more out of date for me over time.

With

```bash
docker run --rm -it -p 8080:8080 zauberzeug/nicegui:1.3.18
```

you can make the docs of a specific version available locally.
